### ISSUE TITLE
[display] incorrect activeParameter in signature help with optional arg skipping

### ISSUE THREAD

**Issue by Gama11** [HaxeFoundation/haxe#7926]
```haxe
class Main {
	static function main() {
		new Main("");
	}

	function new(?a:Int, b:String) {}
}
```

`"activeParameter": 0` is being reported incorrectly here:

![](https://i.imgur.com/VdMeYfl.png)

Instead, 1 should be reported so that the correct argument is highlighted.

**Comment by Simn** [comment]
This requires a bit of a different approach to how we detect this, but I think it's a worthwhile fix to have.

**Comment by Simn** [comment]
Come to think of it, this probably leads to a big mess with overloads. The argument could be number 2 on one overload and number 1 on another.

**Comment by Gama11** [comment]
Yeah, I think I've noticed weird signature help behavior with overloads as well.

**Comment by Simn** [comment]
But what can we do about it?

```haxe
extern class SignatureHelp {
	function new();

	/**
	 * One or more signatures.
	 */
	var signatures:Array<SignatureInformation>;

	/**
	 * The active signature.
	 */
	var activeSignature:Int;

	/**
	 * The active parameter of the active signature.
	 */
	var activeParameter:Int;
}
```

The `activeParameter` is not tied to the list of signatures. So once I switch to a different overload the information of which one is the active parameter is lost.

**Comment by Gama11** [comment]
I think the new signature context API actually added something for that (but that hasn't made it into the LSP yet):

https://github.com/Microsoft/vscode/issues/33413#issuecomment-452096700

**Comment by Simn** [comment]
In that case I prefer to not change this for the time being.

**Comment by Gama11** [comment]
In theory we could always just work around / bypass the LSP, but meh..

**Comment by Simn** [comment]
Do we have what we need in the LSP now?

**Comment by Gama11** [comment]
Yes, I think so: https://github.com/vshaxe/language-server-protocol-haxe/blob/3e21d8ea1dc7b0cff2a615cdd959eab0f2e4f84a/src/languageServerProtocol/protocol/Protocol.hx#L1630
